### PR TITLE
[PeerList][Part 9] Move roundrobin mutex locks from peerring to public roundrobin funcs

### PR DESCRIPTION
**Pull request by willhug**
Summary: In the initial revision the locks for the RoundRobin list were
only around the ring (which was fine).  But with some changes that are
coming (moving peers in lists based on the peer's connection status) we
need to have locks at the top level, so we can move the peerring locks
up one level.

### CODE REVIEW

File: transport/peer/x/peerlist/roundrobin/peerring.go [hunk 1]
```diff
@@ -38,18 +37,14 @@ func NewPeerRing(capacity int) *PeerRing {
...
 type PeerRing struct {
-	lock sync.Mutex
-
 	peerToNode map[string]*peerRingNode
 	nextNode   *peerRingNode
 }
 
 // Add a transport.Peer to the end of the PeerRing, if the ring is empty
 // it initializes the nextNode marker
+// Must be run inside a mutex.Lock()
```
Comment (abhinav) [inline transport/peer/x/peerlist/roundrobin/peerring.go]: That's a weird requirement for a top-level method of a type. There's no
`mutex` in the context of PeerRing.

Instead, remove any references to locks in the comments for PeerRing and
mention in the docstring for PeerRing that this is not thread safe. That
explains to the caller that locking is their concern.

Comment (prashantv) [inline transport/peer/x/peerlist/roundrobin/peerring.go]: +1

File: transport/peer/x/peerlist/roundrobin/roundrobin.go [hunk 2]
```diff
@@ -65,9 +71,12 @@ func (pl *RoundRobin) addAll(peerIDs []transport.PeerIdentifier) error {
 
 // Add a peer identifier to the round robin
 func (pl *RoundRobin) Add(pid transport.PeerIdentifier) error {
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
```
Comment (abhinav) [inline transport/peer/x/peerlist/roundrobin/roundrobin.go]: If there is a single exit path, don't defer. There's some cost associated with
that operation.

```
pl.lock.Lock()
err := pl.addPeer(pid)
pl.lock.Unlock()
return err
```

File: transport/peer/x/peerlist/roundrobin/roundrobin.go [hunk 3]
```diff
@@ -145,8 +160,16 @@ func (pl *RoundRobin) ChoosePeer(ctx context.Context, req *transport.Request) (t
 	}
 }
 
+func (pl *RoundRobin) nextPeer() transport.Peer {
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+
```
Comment (abhinav) [inline transport/peer/x/peerlist/roundrobin/roundrobin.go]: same as above

**Pull request merged**